src/dbnode/storage/index/postings_list_cache.go

```diff
@@ -0,0 +1,253 @@
+	PatternTypeTerm
+
+	reportLoopInterval = 10 * time.Second
+)
+
+// PostingsListCacheOptions is the options struct for the query cache.
+type PostingsListCacheOptions struct {
+	InstrumentOptions instrument.Options
+}
+
+// PostingsListCache implements an LRU for caching queries and their results.
+type PostingsListCache struct {
+	sync.Mutex
+
+	lru *postingsListLRU
+
+	size    int
+	opts    PostingsListCacheOptions
+	metrics *postingsListCacheMetrics
+
+	stopReporting chan struct{}
+}
+
+// NewPostingsListCache creates a new query cache.
+func NewPostingsListCache(size int, opts PostingsListCacheOptions) (*PostingsListCache, error) {
+	lru, err := newPostingsListLRU(size)
+	if err != nil {
+		return nil, err
+	}
+
+	return &PostingsListCache{
+		lru:     lru,
+		size:    size,
+		opts:    opts,
+		metrics: newPostingsListCacheMetrics(opts.InstrumentOptions.MetricsScope()),
+	}, nil
+}
+
+// GetRegexp returns the cached results for the provided regexp query, if any.
+func (q *PostingsListCache) GetRegexp(
+	segmentUUID uuid.UUID,
+	pattern string,
+) (postings.List, bool) {
+	return q.get(
+		segmentUUID,
+		pattern,
+		PatternTypeRegexp)
+}
+
+// GetTerm returns the cached results for the provided term query, if any.
+func (q *PostingsListCache) GetTerm(
+	segmentUUID uuid.UUID,
+	pattern string,
+) (postings.List, bool) {
+	return q.get(
+		segmentUUID,
+		pattern,
+		PatternTypeTerm)
+}
+
+func (q *PostingsListCache) get(
+	segmentUUID uuid.UUID,
+	pattern string,
+	patternType PatternType,
+) (postings.List, bool) {
+	// No RLock because a Get() operation mutates the LRU.
+	q.Lock()
+	p, ok := q.lru.Get(segmentUUID, pattern, patternType)
+	q.Unlock()
+
+	q.emitCacheGetMetrics(patternType, ok)
+
+	if !ok {
+		return nil, false
+	}
+
+	return p, ok
+}
+
+// PutRegexp updates the LRU with the result of the regexp query.
+func (q *PostingsListCache) PutRegexp(
+	segmentUUID uuid.UUID,
+	pattern string,
+	pl postings.List,
+) {
+	q.put(segmentUUID, pattern, PatternTypeRegexp, pl)
+}
+
+// PutTerm updates the LRU with the result of the term query.
+func (q *PostingsListCache) PutTerm(
+	segmentUUID uuid.UUID,
+	pattern string,
+	pl postings.List,
+) {
+	q.put(segmentUUID, pattern, PatternTypeTerm, pl)
+}
+
+func (q *PostingsListCache) put(
+	segmentUUID uuid.UUID,
+	pattern string,
+	patternType PatternType,
+	pl postings.List,
+) {
+	q.Lock()
+	q.lru.Add(
+		segmentUUID,
+		pattern,
+		patternType,
+		pl,
+	)
+	q.Unlock()
+	q.emitCachePutMetrics(patternType)
+}
+
+// PurgeSegment removes all postings lists associated with the specified
+// segment from the cache.
+func (q *PostingsListCache) PurgeSegment(segmentUUID uuid.UUID) {
+	q.Lock()
+	q.lru.PurgeSegment(segmentUUID)
+	q.Unlock()
+}
+
+// StartReportLoop starts a background process that will call Report()
+// on a regular basis and returns a function that will end the background
+// process.
+func (q *PostingsListCache) StartReportLoop() func() {
```

Comment:
Instead of exposing the external report loop controls, how about exposing `Start()` and `Close()` on the cache? Seems more in line with other things we do

Reply:
Yeah it just seems a little weird because there are no resources that need to be released and no background processes going on, so Open()/Close() didn't seem to make a lot of sense.

Reply:
I mean the bg report loop is a resource that needs to be controlled. The current method feels leaky - passing lifecycle hooks outside the function.

Could alternatively do it with a better ctor, something like the Tally Scope idiom.

```
cache, closer, err := index.NewPostingsCache(...)
if err != nil {
...
}
defer closer()
```

Reply:
yeah I like that one. I'll do that